File: LeetCode/1_Two Sum.cpp
// O(n) time complexity, O(n) space
// Hash table  

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> res(2);
        unordered_map<int, int> hashtable;
        for (size_t i = 0; i < nums.size(); i++) {
            if (hashtable.find(target - nums[i]) != hashtable.end()) {
                res[0] = hashtable[target - nums[i]];
                res[1] = i;
                return res;
            }
            else
                hashtable[nums[i]] = i;
        }
    }
};

File: LeetCode/1_Two Sum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        map = {}
        for i, v in enumerate(nums):
            if target - v in map.keys():
                return map[target - v], i
            else:
                map[v] = i